### ISSUE TITLE
Questions about GPR.py

### ISSUE THREAD

**Issue by zhoutianyang2002** [Yoruko-Tang/FedCor#3]
Hi!

I am a newbie of FL. Thank you for your helpful reply last month. Now I am studying your code to learn how to implement a FL experiment. Thank you for your excellent work!

May I ask some questions about the code?

1. I notice that the code provide a function to normalize the poly kernel, and the default option is no normalization. What's the difference or benefits of normalization?

   Besides, in the class Kernel_GPR, when initialize the embedding matrix PMatrix, the code is :

   ```python
   self.PMatrix = Parameter(torch.randn(dimension,num_users)/np.sqrt(dimension))
   ```

   Why we divide by np.sqrt(dimension)) , not just divide by dimension?  Is that makes the ||PMatrix||=1?

2. I am not sure about the utility of $\sigma_f$. The code comment mentioned that this is the diagonal of covariance matrix. However, I notice that this is a scalar, not a vector. How could it represent the diagonal? It seems that the sigma_f is only used in SE_kernel and the normalization of Poly_kernel.

3. The function MLE_Mean calculate the weighted mean of historical data. In the function:

   ```python
   for g in cum_gamma:
       if g == 0.0:
           g+=1e-6
   self.mu = self.mu/cum_gamma
   return self.mu
   ```

   May I ask why add 1e-6, not 1? When we divide by 1e-6, it seems that will make the mu very large.

4. When I running the code, I encounter the warning of "Singular posterior covariance encountered, skip the GPR test in this round!" sometimes. May I ask how could I avoid this? Such as making the $\sigma_n$ more big?

Sorry to bother you. Thank you very much!

**Comment by Yoruko-Tang** [comment]
Hi,

Thank you for your interest in our work. I will try to answer your questions one by one:

1. What is the benefit of normalization?
 * The normalization will provide an additional constraint on the covariance matrix that requires all the clients to have the same variance, which means that it will make the MLE more sample-efficient. But the equal-variance assumption does not always hold and we only use it in the 1SPC setting since each client only has data from one class and thus has similar variance in the loss.
 * The initialization of the PMatrix (i.e., $X$ in the paper) makes the embedding of each client fall on the $\ell_2$-unit sphere, which will help increase the stability of MLE.

2. What is $\sigma_f$?
 * As I mentioned above, if we adopt normalization on the kernel, all the clients will have the same variance (i.e., 1 before we multiply it with $\sigma_f$). And we use $\sigma_f$ as the shared variance for all clients, which is another parameter that can be trained.

3. Why adding $10^{-6}$ to $g$?
 * Notice that if $g=0$, we also have $\mu=0$ and thus a $10^{-6}$ avoids the 0/0 error. You can also add $1$ here since it does not make any difference.

4. How to avoid singular posterior covariance?
 * It is normal to encounter this warning and usually, it does not influence the result. This warning occurs because after adding conditions to get the posterior, the variances of some dimensions become extremely small and thus the covariance matrix becomes singular (This also aligns with Corollary 2 in the paper). Adding $\sigma_n$ can avoid it but will also cause an inaccurate prediction of the posterior loss. Therefore, I suggest to ignore this warning.

**Comment by zhoutianyang2002** [comment]
> Hi,
> 
> Thank you for your interest in our work. I will try to answer your questions one by one:
> 
> 1. What is the benefit of normalization?
> 
> * The normalization will provide an additional constraint on the covariance matrix that requires all the clients to have the same variance, which means that it will make the MLE more sample-efficient. But the equal-variance assumption does not always hold and we only use it in the 1SPC setting since each client only has data from one class and thus has similar variance in the loss.
> * The initialization of the PMatrix (i.e., X in the paper) makes the embedding of each client fall on the ℓ2-unit sphere, which will help increase the stability of MLE.
> 
> 2. What is σf?
> 
> * As I mentioned above, if we adopt normalization on the kernel, all the clients will have the same variance (i.e., 1 before we multiply it with σf). And we use σf as the shared variance for all clients, which is another parameter that can be trained.
> 
> 3. Why adding 10−6 to g?
> 
> * Notice that if g=0, we also have μ=0 and thus a 10−6 avoids the 0/0 error. You can also add 1 here since it does not make any difference.
> 
> 4. How to avoid singular posterior covariance?
> 
> * It is normal to encounter this warning and usually, it does not influence the result. This warning occurs because after adding conditions to get the posterior, the variances of some dimensions become extremely small and thus the covariance matrix becomes singular (This also aligns with Corollary 2 in the paper). Adding σn can avoid it but will also cause an inaccurate prediction of the posterior loss. Therefore, I suggest to ignore this warning.

I understand now. Thank you very much for your clear and helpful reply!

**Comment by AstonLe** [comment]
Hello, i also have a few question in GRP aswell, sorry if the question is so naively because im really new with GP.

**Comment by AstonLe** [comment]
Hello, I also have a few questions about GPR code as well. I apologize if my questions are a bit naive, as im new to both GP and FL.

1.  Firstly, im really confused about the ```gamma```  variable and ```discount``` in your code. Is ```gamma``` the same as $\beta$ in Algorithm 1, or is it the $\gamma$ in Equation 16 in your paper?

2.  What is the $\mathrm{a}$ scale constant in Eq.9 ? I understand how $\gamma$ and $\beta^{\tau}$ work, but for $\mathrm{a}$, how should I initialize it, or is it actually computed elsewhere?

**Comment by Yoruko-Tang** [comment]
Thank you for your questions and sorry for some mismatched notations between the codes and the paper. 

1. ```gamma``` in the code is actually $\theta$ in Appendix C.1 (please refer to "Hyperparameters for FedCor" section). ```discount``` in the code is actually $\beta$ in the paper.

2. Actually what $a$ is does not matter in our method. You can see in Eq. (12) and (13) that $a$ does not influence the result of selection. So you can view it as an arbitrary unknown scale.
